crates/driver/src/domain/competition/solution/fee.rs

```diff
@@ -75,46 +75,83 @@ impl Fulfillment {
+                let current_fee = match protocol_fee {
+                    FeePolicy::Surplus {
+                        factor,
+                        max_volume_factor,
+                    } => self.calculate_fee(
+                        current_order_price_limits.clone(),
+                        prices,
+                        *factor,
+                        *max_volume_factor,
+                    ),
+                    FeePolicy::PriceImprovement {
+                        factor,
+                        max_volume_factor,
+                        quote,
+                    } => {
+                        let price_limits = adjust_quote_to_order_limits(
+                            Order {
+                                sell_amount: current_order_price_limits.sell.0,
+                                buy_amount: current_order_price_limits.buy.0,
+                                side: self.order().side,
+                            },
+                            Quote {
+                                sell_amount: quote.sell.amount.0,
+                                buy_amount: quote.buy.amount.0,
+                                fee_amount: quote.fee.amount.0,
+                            },
+                        )?;
+                        self.calculate_fee(price_limits, prices, *factor, *max_volume_factor)
+                    }
+                    FeePolicy::Volume { factor } => self.fee_from_volume(prices, *factor),
+                }?;
 
-        match self.order().protocol_fees.first() {
-            Some(FeePolicy::Surplus {
-                factor,
-                max_volume_factor,
-            }) => self.calculate_fee(
-                PriceLimits {
-                    sell: self.order().sell.amount,
-                    buy: self.order().buy.amount,
-                },
-                prices,
-                *factor,
-                *max_volume_factor,
-            ),
-            Some(FeePolicy::PriceImprovement {
-                factor,
-                max_volume_factor,
-                quote,
-            }) => {
-                let price_limits = adjust_quote_to_order_limits(
-                    Order {
-                        sell_amount: self.order().sell.amount.0,
-                        buy_amount: self.order().buy.amount.0,
-                        side: self.order().side,
-                    },
-                    Quote {
-                        sell_amount: quote.sell.amount.0,
-                        buy_amount: quote.buy.amount.0,
-                        fee_amount: quote.fee.amount.0,
-                    },
-                )?;
-                self.calculate_fee(price_limits, prices, *factor, *max_volume_factor)
-            }
-            Some(FeePolicy::Volume { factor }) => self.fee_from_volume(prices, *factor),
-            None => Ok(0.into()),
-        }
+                // Recalculate the current price limits in order to apply the next protocol fees
```

Comment:
So we actually have to match on the `side` here?
Shouldn't it be irrelevant whether we increase the sell token price or decrease the buy token price for any given order?